rfcs/20210201-resampling-with-no-pixel-shift/README.md

```diff
@@ -0,0 +1,112 @@
+# RFC: Default ResizeBilinear behavior from TF 1.x in resampling primitive
+
+## Introduction
+
+Default linear algorithm in TF 1.x behaves differently than in version 2.x.
+oneDNN currently supports a version compatible with TF 2.x. Some oneDNN users
+learned their models based on the old and inaccurate version of the linear
+algorithm from TF 1.x. Therefore, they want to use their already learned networks
+based on the old algorithm. However, users which use the old TF 1.x linear algorithm
+and block format are forced to run the reorder primitive every time they want to
+resize an image. This causes a significant slowdown in the performance
+of the entire model if resampling occurs repeatedly.
+
+The difference between the oneDNN results and the default resizeBilinear version is
+based on the assumption of where the pixel center is. oneDNN and TF 2.x assume that
+the center of the pixel is halfway between the pixels. This means that e.g. the top
+left pixel is at the point (0.5; 0.5). This approach produces more accurate results
+than the default ResizeBilinear version. However, this did not mean that there was
+no such option in TensorFlow 1.x. In TF 1.14, the `half_pixel_centers` attribute was
+provided for the ResizeBilinear function. By default, it is set to false. However,
+when set to true, the algorithm produces values according to the version of the
+algorithm available for oneDNN and TF 2.x. Here is a link to the patch that introduced this option:
+https://github.com/tensorflow/tensorflow/commit/3ae2c6691b7c6e0986d97b150c9283e5cc52c15f
+Unfortunately, despite this option, some users still used the old inaccurate version,
+which remained the default version up to TF version 2.x.
+
+The results of the linear algorithm for oneDNN and the different versions of this algorithm 
+in TF are presented below. The following results are based on downsampling the image from
+3x3 to 2x2, where the original image was filled with the values: [[0,1,2], [3,4,5], [6,7,8]]
+
+![](oneDNN_result.png)
+
+![](TF_results.png)
+
+## Proposal
+
+### Option 1 Adding the resampling flags in the descriptor constructor:
+
+``` cpp
+/// Flags for resampling primitive.
+typedef enum {
+    dnnl_resampling_flags_none = 0x0U,
+    dnnl_half_pixel_centers = 0x1U,
+} dnnl_resampling_flags_t;
+```
+
+``` cpp
+/// Flags for resampling primitive.
+enum class resampling_flags : unsigned {
+    none = dnnl_resampling_flags_none,
+    half_pixel_centers = dnnl_half_pixel_centers
+};
+```
+
+``` cpp
+...
+desc(prop_kind aprop_kind, algorithm aalgorithm, resampling_flags flags,
```

Comment:
Flags should be `int` to make possible to pass something like `dnnl_half_pixel_centers | dnnl_something | dnnl_someghing_2`, because the result of this operation is integer value and also the result might not be represented as one of flags.

An example:
```cpp
// main.c
typedef enum {
    foo = 0x1,
    bar = 0x2
} flags_t;

int main() {
    flags_t a = foo & bar; // error
    int b = foo & bar;  // works fine

    return 0;
}

```

```bash
$ g++ main.c
main.c: In function ‘int main()’:
main.c:7:23: error: invalid conversion from ‘int’ to ‘flags_t’ [-fpermissive]
     flags_t a = foo & bar;
```

Reply:
I changed to `unsigned`.